### PR TITLE
Observability Builder Props for Addons

### PR DESCRIPTION
**Pull request by Howlla**
*Issue #, if available:*~

*Description of changes:* Modify some addons to export their Type for props.
Modify observability-builder.ts to accept props for each addon

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by elamaran11 — CHANGES_REQUESTED** [review]
@Howlla Couple of questions for clarity

File: lib/addons/adot/index.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@ import { getAdotCollectorPolicyDocument } from "./iam-policy";
 /**
  * Configuration options for the Adot add-on.
  */
-type AdotCollectorAddOnProps = CoreAddOnProps;
+export type AdotCollectorAddOnProps = Omit<CoreAddOnProps, "saName" | "addOnName" >;
```
[the remark this answers is not in the document]
Comment (Howlla) [reply lib/addons/adot/index.ts:12]: Need to export the props to use in observability builder

File: lib/addons/kube-proxy/index.ts [hunk 2]
```diff
@@ -9,17 +9,27 @@ const versionMap: Map<KubernetesVersion, string> = new Map([
...
+    versionMap: versionMap,
+};
+
 /**
  * Implementation of KubeProxy EKS add-on.
  */
 export class KubeProxyAddOn extends CoreAddOn {
 
-    constructor(version?: string) {
+    constructor(version?: string, props?: kubeProxyAddOnProps) {
```
[the remark this answers is not in the document]
Comment (Howlla) [reply lib/addons/kube-proxy/index.ts:28]: version could be taken in directly from kubeproxyaddon instead.
But previous versions of cdk blueprint took version as first argument so didnt want to modify that.

**Comment by Howlla** [comment]
To support Fargate cluster deployment we need to be able to pass props for coredns and cert-manager.
Those are the 2 addons that need props in the argument. The rest props were exposed for future use cases that may require configuration values to be passed. 

We can remove the `export type` from coredns and adot but then customers would have to define type again locally in their code

**Review by elamaran11 — COMMENTED** [review]
LGTM. @shapirov103 Please check and run `e2e`

**Review by elamaran11 — COMMENTED** [review]
LGTM

**Review by shapirov103 — COMMENTED** [review]
@Howlla definitely a needed change!

The approach is a bit overloaded from the software engineering perspective and does not align well with the builder pattern which is supposed to abstract the complexity to build the object. 

How about this style to enable prop overrides:

```typescript
     ObservabilityBuilder.builder()
            .account(account)
            .region(region)
            .withAddOnProps(ObservabilityBuilder.ADDON.CoreDns, myCoreDnsProps)
            .withAddOnProps(ObservabilityBuilder.ADDON.CertManager, myCertManagerProps)
            .enableNativePatternAddOns()
            .addOns(...addOns)
            .build(scope, stackId);
```

Please let me know if you have any questions.

**Comment by Howlla** [comment]
hi @shapirov103 
Could you help me figure out how to properly type this function?

i tried the following way
```
 public withAddOnProps(addon:typeof addons.HelmAddOn|addons.CoreAddOn, props: addons.CoreAddOnProps | addons.HelmAddOnProps ){
        return this.addOns(
            new addon(props)
        )
    }
   ```
   
   but i get the following error
   ```
   This expression is not constructable.
  Not all constituents of type 'CoreAddOn | typeof HelmAddOn' are constructable.
    Type 'CoreAddOn' has no construct signatures.
    ```

**Comment by shapirov103** [comment]
> hi @shapirov103 Could you help me figure out how to properly type this function?
> 
> i tried the following way
> 
> ```
>  public withAddOnProps(addon:addons.HelmAddOn | addons.CoreAddOn, props: addons.CoreAddOnProps | addons.HelmAddOnProps ){
>         return this.addOns(
>             new addon(props)
>         )
>     }
> ```
> 
> but i get the following error
> 
> ```
> This expression is not constructable.
> Not all constituents of type 'CoreAddOn | HelmAddOn' are constructable.
>  Type 'CoreAddOn' has no construct signatures.
>  ```
> ```

@Howlla I think the easiest approach is going to be something like:

```typescript
class ObservabilityBuilder {
   
   private certManagerProps: CertManagerAddOnProps;
   private awsLoadbalancerProps: AwsLoadBalancerAddOnProps;
 ...

  public withCertManagerProps(props: CertManagerAddOnProps) {
     this.certManagerProps = props; // actually deep clone here to avoid side effects
  }

  private withAwsLoadBalancerProps(props: AwsLoadBalancerAddOnProps) {
    ...
  }
  
 
}


ObservabilityBuilder.builder()
    .withCertManagerProps( {someprops} )
    .withAwsLoadBalancerProps({some props})
    .enable...()

```

The downside is that for each addon there is an override method, but it makes it easy for the users to consume.

**Comment by Howlla** [comment]
> > hi @shapirov103 Could you help me figure out how to properly type this function?
> > i tried the following way
> > ```
> >  public withAddOnProps(addon:addons.HelmAddOn | addons.CoreAddOn, props: addons.CoreAddOnProps | addons.HelmAddOnProps ){
> >         return this.addOns(
> >             new addon(props)
> >         )
> >     }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > but i get the following error
> > ```
> > This expression is not constructable.
> > Not all constituents of type 'CoreAddOn | HelmAddOn' are constructable.
> >  Type 'CoreAddOn' has no construct signatures.
> > ```
> 
> @Howlla I think the easiest approach is going to be something like:
> 
> ```ts
> class ObservabilityBuilder {
>    
>    private certManagerProps: CertManagerAddOnProps;
>    private awsLoadbalancerProps: AwsLoadBalancerAddOnProps;
>  ...
> 
>   public withCertManagerProps(props: CertManagerAddOnProps) {
>      this.certManagerProps = props; // actually deep clone here to avoid side effects
>   }
> 
>   private withAwsLoadBalancerProps(props: AwsLoadBalancerAddOnProps) {
>     ...
>   }
>   
>  
> }
> 
> 
> ObservabilityBuilder.builder()
>     .withCertManagerProps( {someprops} )
>     .withAwsLoadBalancerProps({some props})
>     .enable...()
> ```
> 
> The downside is that for each addon there is an override method, but it makes it easy for the users to consume.

I have made the changes as per your recommendation Mikhail. Will require slight modification to the cdk-eks-observability-accelerator for [these](https://github.com/search?q=repo%3Aaws-observability%2Fcdk-aws-observability-accelerator%20enableOpenSourcePatternAddOns&type=code) files.

**Comment by shapirov103** [comment]
@elamaran11 I put my approval on this, looks like it is pending requested changes from your side, please take a look.

**Pull request merged**